### PR TITLE
Fix MUMPS libraries link sequence

### PR DESCRIPTION
**Pull request by adam-sim-dev**
| PR | Author | Editor | Reviewers | Assignment | Approval | Merge|
 | --- | --- | --- | --- | --- | --- | --- |
| [#4214](https://github.com/mfem/mfem/pull/4214) | @adam-sim-dev | @v-dobrev | @psocratis + @johnbowen42 + @v-dobrev | 3/27/24 | 4/18/24 | 4/18/24 | |

<details>
<summary>PR Checklist</summary>
    
- [ ] Code builds.
- [ ] Code passes `make style`.
- [ ] Update `CHANGELOG`:
    - [ ] Is this a new feature users need to be aware of? New or updated example or miniapp?
    - [ ] Does it make sense to create a new section in the `CHANGELOG` to group with other related features?
- [ ] Update `INSTALL`:
    - [ ] Had a new optional library been added? If so, what range of versions of this library are required? (*Make sure the external library is compatible with our BSD license, e.g. it is not licensed under GPL!*)
    - [ ] Have the version ranges for any required or optional libraries changed?
    - [ ] Does `make` or `cmake` have a new target?
    - [ ] Did the requirements or the installation process change? *(rare)*
- [ ] Update continuous integration server configurations if necessary (e.g. with new version requirements for each of MFEM's dependencies)
    - [ ] `.github`
    - [ ] `.appveyor.yml`
- [ ] Update `.gitignore`:
    - [ ] Check if `make distclean; git status` shows any files that were generated from the source by the project (not an IDE) but we don't want to track in the repository.
    - [ ] Add new patterns (just for the new files above) and re-run the above test.
- [ ] New examples:
    - [ ] All sample runs at the top of the example source file work.
    - [ ] Update `examples/makefile`:
      - [ ] Add the example code to the appropriate `SEQ_EXAMPLES` and `PAR_EXAMPLES` variables.
      - [ ] Add any files generated by it to the `clean` target.
      - [ ] Add the example binary and any files generated by it to the top-level `.gitignore` file.
    - [ ] Update `examples/CMakeLists.txt`:
      - [ ] Add the example code to the `ALL_EXE_SRCS` variable.
      - [ ] Make sure `THIS_TEST_OPTIONS` is set correctly for the new example.
   - [ ] List the new example in `doc/CodeDocumentation.dox`.
   - [ ] If new examples directory (e.g.`examples/pumi`), list it in `doc/CodeDocumentation.conf.in`
   - [ ] Companion pull request for documentation in [mfem/web](https://github.com/mfem/web) repo:
      - [ ] Update or add example-specific documentation, see e.g. the `src/examples.md`.
      - [ ] Add the description, labels and screenshots in `src/examples.md` and `src/img`.
      - [ ] In `examples.md`, list the example under the appropriate categories, add new categories if necessary.
      - [ ] Add a short description of the example in the "Extensive Examples" section of `features.md`.
- [ ] New miniapps:
   - [ ] All sample runs at the top of the miniapp source file work.
   - [ ] Update top-level `makefile` and `makefile` in corresponding miniapp directory.
   - [ ] Add the miniapp binary and any files generated by it to the top-level `.gitignore` file.
   - [ ] Update CMake build system:
      - [ ] Update the `CMakeLists.txt` file in the `miniapps` directory, if the new miniapp is in a new directory.
      - [ ] Add/update the `CMakeLists.txt` file in the new miniapp directory.
      - [ ] Consider adding a new test for the new miniapp.
   - [ ] List the new miniapp in `doc/CodeDocumentation.dox`
   - [ ] If new miniapps directory (e.g.`miniapps/nurbs`), add it to `MINIAPP_SUBDIRS` in the `makefile`.
   - [ ] If new miniapps directory (e.g.`miniapps/nurbs`), list it in `doc/CodeDocumentation.conf.in`
   - [ ] Companion pull request for documentation in [mfem/web](https://github.com/mfem/web) repo:
     - [ ] Update or add miniapp-specific documentation, see e.g. the `src/meshing.md` and `src/electromagnetics.md` files.
     - [ ] Add the description, labels and screenshots in `src/examples.md` and `src/img`.
     - [ ] The miniapps go at the end of the page, and are usually listed only under a specific "Application (PDE)" category.
     - [ ] Add a short description of the miniapp in the "Extensive Examples" section of `features.md`.
- [ ] New capability:
   - [ ] All new public, protected, and private classes, methods, data members, and functions have full Doxygen-style documentation in source comments. Documentation should include descriptions of member data, function arguments and return values, template parameters, and prerequisites for calling new functions.
   - [ ] Pointer arguments and return values must specify whether ownership is being transferred or lent with the call.
   - [ ] Any new functions should include descriptions of their intended use e.g. for internal use only, user-facing, etc., along with references to example code whenever possible/appropriate.
   - [ ] Consider adding new sample runs in existing examples to highlight the new capability.
   - [ ] Consider saving cool simulation pictures with the new capability in the Confluence gallery (LLNL only) or submitting them, via pull request, to the gallery section of the `mfem/web` repo.
   - [ ] If this is a major new feature, consider mentioning it in the short summary inside `README` *(rare)*.
   - [ ] List major new classes in `doc/CodeDocumentation.dox` *(rare)*.
- [ ] Update this checklist, if the new pull request affects it.
- [ ] Run `make unittest` to make sure all unit tests pass.
- [ ] Run the tests in `tests/scripts`.
- [ ] (LLNL only) After merging:
   - [ ] Update internal tests to include the new features.
</details>

### CODE REVIEW

**Comment by v-dobrev** [comment]
This PR is now [under review](https://github.com/mfem/mfem/blob/master/CONTRIBUTING.md#pull-requests) (see the table in the PR description). To help with the review process, please **do not force push** to the branch.

**Review by psocratis — APPROVED** [review]

**Comment by v-dobrev** [comment]
Should we include in this PR the changes mentioned in this comment: https://github.com/mfem/mfem/pull/3922#issuecomment-2023779007?

**Comment by tzanio** [comment]
> Should we include in this PR the changes mentioned in this comment: https://github.com/mfem/mfem/pull/3922#issuecomment-2023779007?

My vote is yes.

**Comment by v-dobrev** [comment]
> Lets make the changes from [#3922 (comment)](https://github.com/mfem/mfem/pull/3922#issuecomment-2023779007) here, @adam-sim-dev can you do that?

@adam-sim-dev, let me know if you need help with this, e.g. the CMake updates.

**Comment by adam-sim-dev** [comment]
> > Lets make the changes from [#3922 (comment)](https://github.com/mfem/mfem/pull/3922#issuecomment-2023779007) here, @adam-sim-dev can you do that?
> 
> @adam-sim-dev, let me know if you need help with this, e.g. the CMake updates.

Which variable (like MFEM_USE_SINGLE) should I use in FindMUMPS.cmake?

File: config/defaults.mk
Comment (v-dobrev) [inline config/defaults.mk:178]: In order for this to work with a user config file, the `?=` need to be replaced with `=`:
```suggestion
   MFEM_USE_DOUBLE = YES
   MFEM_USE_SINGLE = NO
else ifneq ($(filter single Single SINGLE,$(MFEM_PRECISION)),)
   MFEM_USE_DOUBLE = NO
   MFEM_USE_SINGLE = YES
```

**Comment by v-dobrev** [comment]
@johnbowen42, can you help with the CMake modifications here as we discussed earlier?

**Review by v-dobrev — APPROVED** [review]

**Comment by v-dobrev** [comment]
Merged in `next` for testing...

**Comment by v-dobrev** [comment]
@johnbowen42, thank you for the CMake updates!

@psocratis, there were some more changes after you approved -- take another look if you want.

@adam-sim-dev, thank you for the fixes. If you use CMake, can you confirm that @johnbowen42's updates in CMake work for you?

**Comment by v-dobrev** [comment]
By the way, I tested this PR with both GNU make and CMake and both single and double precision -- everything seems to work as expected.

**Review by psocratis — APPROVED** [review]

**Review by johnbowen42 — APPROVED** [review]
LGTM

**Comment by v-dobrev** [comment]
Re-merged in `next` for testing...

**Pull request merged**
